src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs

```diff
@@ -322,6 +324,299 @@ public virtual void Encode(TextWriter output, char[] value, int startIndex, int
             }
         }
 
+        public unsafe virtual OperationStatus Encode(ReadOnlySpan<char> source, Span<char> destination, out int charsConsumed, out int charsWritten, bool isFinalBlock = true)
+        {
+            // Optimization: Detect how much "doesn't require escaping" data exists at the beginning of the buffer,
+            // and memcpy it directly to the destination.
+
+            int idx = FindFirstCharacterToEncode(source);
```

Comment:
The overhead should be minimal. If the caller follows the general pattern of `FindFirstChar` / (manually copy up to this point) / `Encode`, then the subsequent call to `FindFirstChar` will return immediately because it'll match on the first byte of the input. Alternatively, the caller could skip the call to `FindFirstChar` entirely, and the method will perform the copy on the caller's behalf.